### ISSUE TITLE
@storageRestrictions(initializes:) rejects usage of `self`

### ISSUE THREAD

**Issue by MahdiBM** [apple/swift#74075]
### Description

`@storageRestrictions(initializes: self.__myVar)` in init-accessors is rejected by the compiler, but `@storageRestrictions(initializes: __myVar)` is accepted. I expect the compiler to be fine with the explicit `self`.

Example that works:
```swift
struct MyStruct {
    private var __myVar: String
    var myVar: String {
        @storageRestrictions(initializes: __myVar)
        init {
            self.__myVar = newValue /// Original code makes some adjustments here
        }
        set {
            self.__myVar = newValue /// Original code makes some adjustments here
        }
        get {
            self.__myVar
        }
    }
}
```
This doesn't work:
```diff
-        @storageRestrictions(initializes: __myVar)
+        @storageRestrictions(initializes: self.__myVar)
```

### Reproduction

Described above.

### Expected behavior

`@storageRestrictions(initializes: self.__myVar)` which has explicit `self`, should be accepted.

### Environment

```
swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0
```

### Additional information

A quick skim over the [proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0400-init-accessors.md), it doesn't seem to have opinions about this.
Another quick skim over the [review post](https://forums.swift.org/t/se-0400-init-accessors/65583), I don't see any mentions of this there either.

**Comment by ahoppen** [comment]
This behaves correctly. The `@storageRestrictions` attribute takes a list of member names, not expressions, so `self.__myVar` is not supported.

**Comment by MahdiBM** [comment]
Hmm i see where you're coming from, but it still _feels_ weird 🤔

**Comment by AnthonyLatsis** [comment]
Quoting the proposal just to reassert that the behavior is correct:

> ```
> storage-restrictions-initializes ::= 'initializes' ':' identifier-list
> storage-restrictions-accesses ::= 'accesses' ':' identifier-list
> ```

But I would argue that this case deserves a tailored error and a fix-it — #74190.
